File: shared/web-sdk/src/connectors/sessions-connector.test.ts
import { describe, expect, it, vi } from "vitest";
import type { SessionsListResult, SessionsRest } from "../sessions-rest.ts";
import { SessionsConnector } from "./sessions-connector.ts";

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

const fakeSdk = () => {
  const handlers: Record<string, ((m: unknown) => void)[]> = {};
  const sent: Record<string, unknown>[] = [];
  return {
    onMessage(type: string, fn: (m: unknown) => void) {
      const list = handlers[type] ?? [];
      list.push(fn);
      handlers[type] = list;
      return () => {};
    },
    send(msg: Record<string, unknown>) {
      sent.push(msg);
    },
    emit(type: string, msg: unknown) {
      for (const h of handlers[type] ?? []) h(msg);
    },
    sent,
    sendBinary: vi.fn(),
    onBinary: vi.fn(() => () => {}),
  };
};

const makeRow = (id: string) => ({
  sessionId: id,
  rootId: id,
  title: "T",
  startedAt: 1,
  lastActiveAt: 1,
  messageCount: 0,
  isActive: false,
});

const makeRest = (overrides: Partial<SessionsRest> = {}): SessionsRest => ({
  list: vi.fn(async () => ({ items: [], total: 0, hasMore: false }) as SessionsListResult),
  search: vi.fn(async () => []),
  getMessages: vi.fn(async () => []),
  rename: vi.fn(async () => {}),
  delete: vi.fn(async () => {}),
  ...overrides,
});

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe("SessionsConnector", () => {
  it("list — delegates to REST client with default opts", async () => {
    const sdk = fakeSdk();
    const rest = makeRest({
      list: vi.fn(async () => ({ items: [makeRow("s-1")], total: 1, hasMore: false })),
    });
    const c = new SessionsConnector({ rest });
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    const result = await c.list();
    expect(result.items[0]?.sessionId).toBe("s-1");
    expect(rest.list).toHaveBeenCalledWith({});
  });

  it("list — passes limit and offset to REST", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    await c.list({ limit: 20, offset: 10 });
    expect(rest.list).toHaveBeenCalledWith({ limit: 20, offset: 10 });
  });

  it("search — delegates to REST client", async () => {
    const sdk = fakeSdk();
    const row = makeRow("s-2");
    const rest = makeRest({ search: vi.fn(async () => [row]) });
    const c = new SessionsConnector({ rest });
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    const items = await c.search("hello", 10);
    expect(items).toEqual([row]);
    expect(rest.search).toHaveBeenCalledWith("hello", 10);
  });

  it("delete — calls REST and dispatches deleted change event", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    const onChange = vi.fn();
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    c.onSessionsChanged(onChange);
    await c.delete("s-1");
    expect(rest.delete).toHaveBeenCalledWith("s-1");
    expect(onChange).toHaveBeenCalledWith({ kind: "deleted", sessionId: "s-1" });
  });

  it("rename — calls REST and dispatches renamed change event", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    const onChange = vi.fn();
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    c.onSessionsChanged(onChange);
    await c.rename("s-1", "New Title");
    expect(rest.rename).toHaveBeenCalledWith("s-1", "New Title");
    expect(onChange).toHaveBeenCalledWith({ kind: "renamed", sessionId: "s-1", title: "New Title" });
  });

  it("delete — does not dispatch change event if REST rejects", async () => {
    const sdk = fakeSdk();
    const rest = makeRest({
      delete: vi.fn(async () => {
        throw new Error("not found");
      }),
    });
    const c = new SessionsConnector({ rest });
    const onChange = vi.fn();
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    c.onSessionsChanged(onChange);
    await expect(c.delete("s-1")).rejects.toThrow("not found");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("switchTo — sends conversation.activate (not session.switch) over WS", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    const p = c.switchTo("s-1");
    // Must send conversation.activate, not session.switch
    expect(sdk.sent[0]).toMatchObject({ type: "conversation.activate", sessionId: "s-1" });
    expect(sdk.sent[0]).not.toMatchObject({ type: "session.switch" });
    sdk.emit("session.switched", { sessionId: "s-1", ts: 1 });
    await expect(p).resolves.toBeUndefined();
  });

  it("switchTo — dispatches switched change event on session.switched", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    const onChange = vi.fn();
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    c.onSessionsChanged(onChange);
    const p = c.switchTo("s-1");
    sdk.emit("session.switched", { sessionId: "s-1", ts: 42 });
    await p;
    expect(onChange).toHaveBeenCalledWith({ kind: "switched", sessionId: "s-1", ts: 42 });
  });

  it("newChat — resolves with sessionId on session.created", async () => {
    const sdk = fakeSdk();
    const rest = makeRest();
    const c = new SessionsConnector({ rest });
    c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
    const p = c.newChat();
    sdk.emit("session.created", { sessionId: "s-new", ts: 1 });
    await expect(p).resolves.toEqual({ sessionId: "s-new" });
  });

  it("switchTo — rejects on timeout", async () => {
    vi.useFakeTimers();
    try {
      const sdk = fakeSdk();
      const rest = makeRest();
      const c = new SessionsConnector({ rest, timeoutMs: 50 });
      c.attach(sdk as unknown as Parameters<typeof c.attach>[0]);
      const p = c.switchTo("s-1");
      vi.advanceTimersByTime(60);
      await expect(p).rejects.toThrow(/timeout/i);
    } finally {
      vi.useRealTimers();
    }
  });
});

File: shared/mobile-sdk/src/commonTest/kotlin/io/sentient/mobilesdk/connectors/SessionsConnectorTest.kt
// ---------------------------------------------------------------------------
// SessionsConnectorTest — pins the session lifecycle frame contract and REST
// delegation after protocol Task 2.1/2.6.
//
// Query RPCs (list/search/delete/rename) now delegate to SessionsHttpClient
// (REST). The WS connector only owns lifecycle frames:
//   switchTo   → sends conversation.activate; resolves on session.switched
//   newChat    → sends session.new; resolves on session.created
//
// Wire/protocol contract → keeper per .claude/rules/testing.md:
//   - switchTo sends conversation.activate (not session.switch)
//   - newChat still sends session.new
//   - session.switched resolves switchTo; session.created resolves newChat
//   - broadcasts fan out to onSessionsChanged
//   - timeout on switch throws SessionsTimeoutException
//   - reset() fails pending waiters
//   - list/delete/rename delegate to httpClient (no WS frame sent)
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.http.HttpStatusCode
import io.sentient.mobilesdk.protocol.ClientMessage
import io.sentient.mobilesdk.protocol.ConversationFeedItem
import io.sentient.mobilesdk.protocol.ServerMessage
import io.sentient.mobilesdk.protocol.SessionRow
import io.sentient.mobilesdk.sessions.SessionsHttpClient
import io.sentient.mobilesdk.util.Clock
import kotlinx.coroutines.async
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertTrue

// ── Test helpers ──────────────────────────────────────────────────────────────

private fun row(id: String) = SessionRow(
    sessionId = id,
    rootId = null,
    title = "t-$id",
    startedAt = 0L,
    lastActiveAt = 0L,
    messageCount = 0,
    isActive = false,
)

private fun counterIds(): () -> String {
    var n = 0
    return { "r${n++}" }
}

/**
 * Minimal fake REST client. Uses a stub MockEngine (never called in unit tests
 * since we directly call the fake methods via FakeSessionsHttpClient).
 */
private class FakeSessionsHttpClient(
    private val deleteSuccess: Boolean = true,
    private val renameSuccess: Boolean = true,
) : SessionsHttpClient(
    httpClient = HttpClient(MockEngine { _ -> respond("", HttpStatusCode.OK) }),
    gatewayWsUrl = "wss://h/api/v1/ws",
    token = { "" },
) {
    val listCalls = mutableListOf<Pair<Int, Int>>()
    val deleteCalls = mutableListOf<String>()
    val renameCalls = mutableListOf<Pair<String, String>>()
    var listResult: List<SessionRow> = emptyList()

    override suspend fun list(limit: Int, offset: Int): List<SessionRow> {
        listCalls += limit to offset
        return listResult
    }

    override suspend fun delete(sessionId: String): Boolean {
        deleteCalls += sessionId
        return deleteSuccess
    }

    override suspend fun rename(sessionId: String, title: String): Boolean {
        renameCalls += sessionId to title
        return renameSuccess
    }

    override suspend fun getMessages(sessionId: String, limit: Int, offset: Int): List<ConversationFeedItem> =
        emptyList()

    override suspend fun search(q: String, limit: Int): List<SessionRow> = emptyList()
}

private fun connector(
    sent: MutableList<ClientMessage>,
    httpClient: SessionsHttpClient? = null,
    timeoutMs: Long = 5_000L,
): SessionsConnector =
    SessionsConnector(
        send = { sent += it },
        newId = counterIds(),
        clock = Clock { 0L },
        httpClient = httpClient,
        timeoutMs = timeoutMs,
    )

class SessionsConnectorTest {

    @Test
    fun has_capability_sessions() {
        assertEquals("sessions", connector(mutableListOf()).capability)
    }

    // ── switchTo sends conversation.activate ──────────────────────────────────

    @Test
    fun switchTo_sends_conversation_activate() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val c = connector(sent)
        val deferred = async { c.switchTo("s1") }
        runCurrent()

        val frame = sent.single()
        assertIs<ClientMessage.ConversationActivate>(frame)
        assertEquals("s1", frame.sessionId)
        deferred.cancel()
    }

    @Test
    fun switchTo_resolves_on_session_switched_broadcast() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val c = connector(sent)
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }
        val deferred = async { c.switchTo("s1") }
        runCurrent()

        c.handle(ServerMessage.SessionSwitched(sessionId = "s1", ts = 1L))

        deferred.await()
        assertEquals(listOf<SessionsChangeEvent>(SessionsChangeEvent.Switched("s1", null, 1L)), events)
    }

    @Test
    fun switchTo_times_out_and_throws_typed_exception() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val c = connector(sent, timeoutMs = 50L)
        val captured = async { runCatching { c.switchTo("s1") } }
        runCurrent()
        advanceTimeBy(60L)
        runCurrent()

        val ex = captured.await().exceptionOrNull()
        assertTrue(ex is SessionsTimeoutException)
        assertEquals("session.switched", ex.frameType)
    }

    // ── newChat sends session.new ──────────────────────────────────────────────

    @Test
    fun newChat_sends_session_new() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val c = connector(sent)
        val deferred = async { c.newChat() }
        runCurrent()

        assertIs<ClientMessage.SessionNew>(sent.single())
        deferred.cancel()
    }

    @Test
    fun newChat_resolves_with_sessionId_on_session_created() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val c = connector(sent)
        val deferred = async { c.newChat() }
        runCurrent()

        c.handle(ServerMessage.SessionCreated(sessionId = "s2", ts = 1L))
        assertEquals("s2", deferred.await())
    }

    // ── broadcast fan-out ────────────────────────────────────────────────────

    @Test
    fun deleted_broadcast_fans_out_to_listeners() {
        val c = connector(mutableListOf())
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.handle(ServerMessage.SessionsDeleted(sessionId = "s1"))

        assertEquals(listOf<SessionsChangeEvent>(SessionsChangeEvent.Deleted("s1")), events)
    }

    @Test
    fun renamed_broadcast_fans_out_to_listeners() {
        val c = connector(mutableListOf())
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.handle(ServerMessage.SessionsRenamed(sessionId = "s1", title = "Renamed"))

        assertEquals(listOf<SessionsChangeEvent>(SessionsChangeEvent.Renamed("s1", "Renamed")), events)
    }

    // ── reset fails in-flight waiters ─────────────────────────────────────────

    @Test
    fun reset_fails_in_flight_switch_waiter() = runTest {
        val c = connector(mutableListOf())
        val captured = async { runCatching { c.switchTo("s1") } }
        runCurrent()

        c.reset()

        assertTrue(captured.await().exceptionOrNull() is SessionsTimeoutException)
    }

    @Test
    fun reset_fails_in_flight_newChat_waiter() = runTest {
        val c = connector(mutableListOf())
        val captured = async { runCatching { c.newChat() } }
        runCurrent()

        c.reset()

        assertTrue(captured.await().exceptionOrNull() is SessionsTimeoutException)
    }

    // ── REST delegation: no WS frames sent ───────────────────────────────────

    @Test
    fun list_delegates_to_httpClient_and_sends_no_ws_frame() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val http = FakeSessionsHttpClient().also { it.listResult = listOf(row("s1")) }
        val c = connector(sent, httpClient = http)

        val page = c.list(limit = 20, offset = 0)

        assertEquals(listOf(row("s1")), page.items)
        assertTrue(sent.isEmpty(), "list must NOT send WS frame, sent=$sent")
        assertEquals(listOf(20 to 0), http.listCalls)
    }

    @Test
    fun delete_delegates_to_httpClient_and_fans_out() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val http = FakeSessionsHttpClient(deleteSuccess = true)
        val c = connector(sent, httpClient = http)
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.delete("s-99")

        assertEquals(listOf("s-99"), http.deleteCalls)
        assertEquals(listOf<SessionsChangeEvent>(SessionsChangeEvent.Deleted("s-99")), events)
        assertTrue(sent.isEmpty(), "delete must NOT send WS frame, sent=$sent")
    }

    @Test
    fun delete_does_not_fan_out_on_rest_failure() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val http = FakeSessionsHttpClient(deleteSuccess = false)
        val c = connector(sent, httpClient = http)
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.delete("s-99")

        assertEquals(listOf("s-99"), http.deleteCalls, "REST call must still be attempted")
        assertEquals(emptyList(), events, "no fan-out when REST call fails")
        assertTrue(sent.isEmpty(), "delete must NOT send WS frame, sent=$sent")
    }

    @Test
    fun rename_delegates_to_httpClient_and_fans_out() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val http = FakeSessionsHttpClient(renameSuccess = true)
        val c = connector(sent, httpClient = http)
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.rename("s-5", "My Chat")

        assertEquals(listOf("s-5" to "My Chat"), http.renameCalls)
        assertEquals(listOf<SessionsChangeEvent>(SessionsChangeEvent.Renamed("s-5", "My Chat")), events)
        assertTrue(sent.isEmpty(), "rename must NOT send WS frame, sent=$sent")
    }

    @Test
    fun rename_does_not_fan_out_on_rest_failure() = runTest {
        val sent = mutableListOf<ClientMessage>()
        val http = FakeSessionsHttpClient(renameSuccess = false)
        val c = connector(sent, httpClient = http)
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.rename("s-5", "My Chat")

        assertEquals(listOf("s-5" to "My Chat"), http.renameCalls, "REST call must still be attempted")
        assertEquals(emptyList(), events, "no fan-out when REST call fails")
        assertTrue(sent.isEmpty(), "rename must NOT send WS frame, sent=$sent")
    }

    @Test
    fun ignores_unowned_frames() {
        val c = connector(mutableListOf())
        val events = mutableListOf<SessionsChangeEvent>()
        c.onSessionsChanged { events += it }

        c.handle(ServerMessage.Pong)

        assertEquals(emptyList(), events)
    }
}
